src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/entity/CsipRecord.kt

```diff
@@ -128,5 +134,67 @@ data class CsipRecord(
     this.referral = referral
   }
 
-  fun <T : DomainEvent<AdditionalInformation>> registerEntityEvent(entityEvent: BaseEntityEvent<T>) = apply { registerEvent(entityEvent) }
+  fun <T : DomainEvent<AdditionalInformation>> registerEntityEvent(entityEvent: BaseEntityEvent<T>) =
+    apply { registerEvent(entityEvent) }
+
+  fun registerCsipEvent(domainEvent: CsipEvent) = apply { registerEvent(domainEvent) }
+
+  fun create(
+    createCsipRecordRequest: CreateCsipRecordRequest,
+    csipRequestContext: CsipRequestContext,
+    incidentType: ReferenceData,
+    incidentLocation: ReferenceData,
+    referrerAreaOfWork: ReferenceData,
+    incidentInvolvement: ReferenceData,
+    contributoryFactors: Map<String, ReferenceData>,
+    reason: Reason = Reason.USER,
+    description: String = DomainEventType.CSIP_CREATED.description,
+  ): CsipRecord = apply {
+    referral = createCsipRecordRequest.toInitialReferralEntity(
+      this,
+      csipRequestContext,
+      incidentType,
+      incidentLocation,
+      referrerAreaOfWork,
+      incidentInvolvement,
+    ).apply {
+      createCsipRecordRequest.referral.contributoryFactors.forEach { factor ->
+        this.addContributoryFactor(
+          createRequest = factor,
+          factorType = contributoryFactors[factor.factorTypeCode]!!,
+          actionedAt = createdAt,
+          actionedBy = createdBy,
+          actionedByDisplayName = createdByDisplayName,
+          source = csipRequestContext.source,
+        )
+      }
+    }
+    addAuditEvent(
+      action = AuditEventAction.CREATED,
+      description = "CSIP record created via referral with ${referral!!.contributoryFactors().size} contributory factors",
+      actionedAt = createdAt,
+      actionedBy = createdBy,
+      actionedByCapturedName = createdByDisplayName,
+      source = csipRequestContext.source,
+      reason = reason,
+      activeCaseLoadId = csipRequestContext.activeCaseLoadId,
+      isRecordAffected = true,
+      isReferralAffected = referral != null,
```

Comment:
(Minor) we know that the referral can't be null at this stage (referral!! above and indeed from the toInitialReferralEntity) so this can simply be true